File: 17.letter-combinations-of-a-phone-number.java
/*
 * @lc app=leetcode id=17 lang=java
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start

import java.util.*;

class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result= new ArrayList<>();
        if(digits==null || digits.length()==0){
            return result;
        
        } 
        String[] map ={
            "","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"
        };
        backtrack(0,digits,new StringBuilder(),result,map);
        return result;
        
    }
    private void backtrack(int index,String digits,StringBuilder current,List<String> result,String [] map){
        //base cse
        if(index==digits.length()){
            result.add(current.toString());
            return;
        }
        String letters =map[digits.charAt(index)-'0'];
        for(char ch : letters.toCharArray()){
            current.append(ch);
            backtrack(index+1, digits, current, result, map);
            current.deleteCharAt(current.length()-1);
        }
    }
}
// @lc code=end

File: 17.letter-combinations-of-a-phone-number.py
#
# @lc app=leetcode id=17 lang=python
#
# [17] Letter Combinations of a Phone Number
#

# @lc code=start
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        phone={
            "2":"abc",
            "3":"def",
            "4":"ghi",
            "5":"jkl",
            "6":"mno",
            "7":"pqrs",
            "8":"tuv",
            "9":"wxyz"
        }
        result=[]
        def backtrack(index,path):
            if index==len(digits):
                result.append(path)
                return
            letters=phone[digits[index]]
            for letter in letters:
                backtrack(index+1,path+letter)
        backtrack(0,"")
        return result
        
# @lc code=end

